yacl/crypto/primitives/tpre/tpre.h

```diff
@@ -0,0 +1,124 @@
+#define YACL_CRYPTO_PRIMITIVES_TPRE_TPRE_H_
+
+#include <iostream>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "yacl/crypto/primitives/tpre/capsule.h"
+#include "yacl/crypto/primitives/tpre/keys.h"
+
+namespace yacl::crypto {
+/**
+ * The TPRE class encapsulates the methods of threshold proxy
+ * re-encryption, including encryption, decryption, re-encryption and
+ * multi-party decryption
+ *
+ * Here, let's assume that there are N+2 participants, including Alice, the data
+ * holder, Bob, the data requester, and multiple proxy re-encryption nodes
+ * P_0,..., P_{N-1}.
+ *
+ * Assume m is a data to shared. Alice and Bob will perform the following
+ * calculation steps:
+ *
+ * 1. Alice generates the ciphertext c of plaintext m via
+ *    Encrypt(pk_A, m,...)；Here，Alice can decryt c to m by Decrypt(sk_A,
+ *    c,...)
+ *
+ * 2. Alice entrusts P_1,..., P_t re-encrypts c to obtain capsule set =
+ *    {cfrag_1,...,cfrag_t}. Only t Proxys participate here, reflecting the idea
+ *    of threshold. In the re-encryption phase, you need to use the
+ *    re-encryption key The re-encryption key is generated by calling the
+ *    re-encryption key generation algorithm Keys::GenerateReKey by entering
+ *    Alice's private key and Bob's public key.
+ *
+ * 3. Inputs capsule set and secret key of Bob, decrypts the capsule set to
+ *    plaintext.
+ *
+ * For the description of capsule, you can see in the Capsule class.
+ */
+class TPRE {
+ public:
+  TPRE(){};
+  ~TPRE(){};
+
+  //   /// @brief Rename the re-encryption ciphertext set,  which contains
+  //   capsule
+  //   /// fragments set(CFrags set) and enc_data
+  //   typedef std::pair<std::vector<std::unique_ptr<Capsule::CFrag>>,
+  //                     std::vector<uint8_t>>
+  //       std::pair<std::vector<std::unique_ptr<Capsule::CFrag>>,
+  //                     std::vector<uint8_t>>;
+
+  /// @brief TPRE encryption algorithm
+  /// @param ecc_group, Algebraic operations are on elliptic curves
+  /// @param pk_A, the public key of Alice, Alice refers to the authorizer
+  /// @param iv, random value
+  /// @param plaintext
+  /// @return capsule and encrypted data
+  std::pair<std::unique_ptr<Capsule::CapsuleStruct>, std::vector<uint8_t>>
+  Encrypt(std::unique_ptr<EcGroup> ecc_group,
+          std::unique_ptr<Keys::PublicKey> pk_A, const std::string& iv,
+          const std::string& plaintext);
+
+  /// @brief TPRE decryption algorithm
+  /// @param ecc_group, Algebraic operations are on elliptic curves
+  /// @param capsule_struct, reference Capsule class.
+  /// @param iv, random value
+  /// @param enc_data, decrypted data
+  /// @param sk_A, the secret key of Alice, Alice refers to the authorizer
+  /// @return plaintext
+  std::string Decrypt(std::unique_ptr<EcGroup> ecc_group,
+                      std::unique_ptr<Capsule::CapsuleStruct> capsule_struct,
+                      const std::string& iv, std::vector<uint8_t> enc_data,
```

Comment:
P? pass parameter by const reference
```suggestion
                      const std::string& iv, const std::vector<uint8_t>& enc_data,
```

Reply:
Revised according to comments